torch/csrc/api/include/torch/python.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <Python.h>
+
+#include <pybind11/pybind11.h>
+#include <torch/csrc/utils/pybind.h>
+#include <torch/csrc/variable_tensor_functions.h>
+
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace python {
+namespace detail {
+template <typename Cursor>
+std::vector<Tensor> cursor_to_vector(const Cursor& cursor) {
+  std::vector<Tensor> vector;
+  vector.reserve(cursor.size());
+  cursor.map(
+      std::back_inserter(vector), [](const Tensor& tensor) { return tensor; });
+  return vector;
+}
+
+template <typename Cursor>
+std::unordered_map<std::string, Tensor> cursor_to_map(const Cursor& cursor) {
+  std::unordered_map<std::string, Tensor> map;
+  map.reserve(cursor.size());
+  cursor.map_items(
+      std::inserter(map, map.end()),
+      [](const std::string& key, const Tensor& tensor) {
+        return std::make_pair(key, tensor);
+      });
+  return map;
+}
+} // namespace detail
+
+/// Adds method bindings for a pybind11 `class_` that binds an `nn::Module`
+/// subclass.
+///
+/// Say you have a pybind11 class object created with `py::class_<Net>(m,
+/// "Net")`. This function will add all the necessary `.def()` calls to bind the
+/// `nn::Module` base class' methods, such as `train()`, `eval()` etc. into
+/// Python. The exact list of supported methods and their Python signatures are:
+/// - `train()`
+/// - `eval()`
+/// - `is_training() -> bool`
+/// - `zero_grad()`
+/// - `cuda()`
+/// - `cpu()`
+/// - `parameters() -> List<Tensor>`
+/// - `named_parameters() -> Dict<String, Tensor>`
+/// - `buffers() -> List<Tensor>`
+/// - `named_buffers() -> Dict<String, Tensor>`
+template <typename M, typename... Extra>
+py::class_<M, Extra...> add_module_bindings(py::class_<M, Extra...> module) {
+  return module.def("train", [](M& module) { module.train(); })
+      .def("eval", [](M& module) { module.eval(); })
+      .def("is_training", [](M& module) { return module.is_training(); })
+      .def("zero_grad", [](M& module) { module.zero_grad(); })
+      .def("cuda", [](M& module) { module.to(torch::kCUDA); })
+      .def("cpu", [](M& module) { module.to(torch::kCPU); })
```

Comment:
Could we also bind the `.to` method to python? This is the recommended way now I believe

Reply:
I have to check if a `torch.device` will convert into the right thing on Python side, otherwise yes good idea